graylog2-web-interface/src/views/logic/valueactions/createEventDefinition/hooks/hookHelpers.ts

```diff
@@ -0,0 +1,182 @@
+
+    return res;
+  }, new Set([]));
+};
+
+export const filtratePathsByPivot = ({ flattenPivots, valuePath }: {flattenPivots: Set<string>, valuePath: Array<{ [name: string]: string }>}): Array<{[name:string]: string}> => {
+  if (!valuePath) return [];
+  const map = valuePath.reduce((res, cur) => {
+    const key = Object.keys(cur)[0];
+
+    if (!res.has(key) && flattenPivots.has(key)) {
+      res.set(key, cur);
+    }
+
+    return res;
+  }, new Map([])) as Map<string, {[name:string]: string}>;
+
+  return Array.from(map.values());
+};
+
+export const aggregationMetricValueHandler: AggregationHandler = ({ widget, value, field, valuePath }) => {
+  const curSeries = widget.config.series.find((series) => series.function === field);
+  const { field: agg_field, function: agg_function } = seriesToMetrics([curSeries])[0];
+  const { rowPivots, columnPivots } = widget.config;
+  const flattenRowPivots = getFlattenPivots(rowPivots);
+  const flattenColumnPivots = getFlattenPivots(columnPivots);
+  const rowPaths = filtratePathsByPivot({ flattenPivots: flattenRowPivots, valuePath });
+  const columnPaths = filtratePathsByPivot({ flattenPivots: flattenColumnPivots, valuePath });
+  const rowValuePath = transformValuePathToQuery(rowPaths);
+  const columnValuePath = transformValuePathToQuery(columnPaths);
+
+  return ({
+    aggField: agg_field,
+    aggFunction: agg_function,
+    aggValue: value,
+    rowGroupBy: Array.from(flattenRowPivots),
+    columnGroupBy: Array.from(flattenColumnPivots),
+    rowValuePath,
+    columnValuePath,
+  });
+};
+
+export const aggregationValueHandler: AggregationHandler = ({ widget, value, field, valuePath }) => {
+  const { rowPivots } = widget.config;
+  const flattenRowPivots = getFlattenPivots(rowPivots);
+  const rowPaths = filtratePathsByPivot({ flattenPivots: flattenRowPivots, valuePath });
+  const rowValuePath = transformValuePathToQuery(rowPaths);
+
+  return ({
+    searchFromValue: `${field}:${value}`,
+    rowValuePath,
+  });
+};
+
+export const messagesValueHandler: AggregationHandler = ({ value, field }) => {
+  return ({
+    searchFromValue: `${field}:${value}`,
+  });
+};
+
+export const logsValueHandler: AggregationHandler = ({ value, field }) => {
+  return ({
+    searchFromValue: `${field}:${value}`,
+  });
+};
+
+export const getAggregationHandler = ({ widget, field }: { widget: Widget, field: string }): AggregationHandler => {
+  if (widget.type === 'AGGREGATION') {
+    const isMetrics = !!widget.config.series.find((series) => series.function === field);
+
+    return isMetrics ? aggregationMetricValueHandler : aggregationValueHandler;
+  }
+
+  if (widget.type === 'logs') return logsValueHandler;
+  if (widget.type === 'messages') return messagesValueHandler;
+
+  throw new Error('This widget type has incorrect type or no handler');
+};
+
+export const getLutParameters = (parameters: Immutable.Set<Parameter>) => parameters.reduce((res, cur) => {
```

Comment:
Why do we need the `toJSON()` here?

Reply:
> Why do we need the `toJSON()` here?

@dennisoelkers Later it will be transformed as a part of a stringified object and placed to the URL. In this case, would be better to have just an object.